src/Orleans/Async/AsyncExecutorWithRetries.cs

```diff
@@ -116,72 +116,87 @@ internal static class AsyncExecutorWithRetries
 
             do
             {
-                retry = false;
-
-                if (maxExecutionTime != Constants.INFINITE_TIMESPAN && maxExecutionTime != default(TimeSpan))
-                {
-                    DateTime now = DateTime.UtcNow;
-                    if (now - startExecutionTime > maxExecutionTime)
-                    {
-                        throw new TimeoutException(
-                            $"ExecuteWithRetries has exceeded its max execution time of {maxExecutionTime}. Now is {LogFormatter.PrintDate(now)}, started at {LogFormatter.PrintDate(startExecutionTime)}, passed {now - startExecutionTime}");
-                    }
-                }
-
                 int counter = callCounter;
 
                 try
                 {
                     callCounter++;
-                    result = await function(counter);
+                    result = await function(counter).WithTimeout(GetTimeout(startExecutionTime, maxExecutionTime));
```

Comment:
I understand this change and see the value in it, but I'm still a bit concerned by it.  It is a behavioral change in which successful operations can be returned as timeout errors.  It is an improvement in the callers control of how much time is spent retrying, but at a cost.

As the current timeout behavior has not demonstrated itself to be flawed or insufficient, I submit that we should stick with it until we've a need to change it.